tests/unit/test_spmd.c

```diff
@@ -0,0 +1,254 @@
+        sdispls[i] = (((i+1) * (i))/2) * sizeof(int);
+        sendtypes[i] = recvtypes[i] = MPI_INT;
+    }
+
+    /* Free memory. */
+    free(sendcounts);
+    free(recvcounts);
+    free(rdispls);
+    free(sdispls);
+    free(sendtypes);
+    free(recvtypes);
+
+
+    //    for (int msg_cnt=4; msg_cnt<size; msg_cnt*=2){
+    //   if (rank==0) printf("message count %d\n",msg_cnt);
+    int msg_cnt = 0;
+    for (int itest = 0; itest < 5; itest++)
+    {
+        bool hs = false;
+        bool isend = false;
+
+        /* Load up the buffers */
+        for (int i = 0; i < num_elem; i++)
+        {
+            sbuf[i] = i + 100 * my_rank;
+            rbuf[i] = -i;
+        }
+        MPI_Barrier(test_comm);
+
+        if (!my_rank)
+        {
+            printf("Start itest %d\n", itest);
+            gettimeofday(&t1, NULL);
+        }
+
+        if (itest == 0)
+            ret = pio_swapm(sbuf, sendcounts, sdispls, sendtypes, rbuf, recvcounts,
+                            rdispls, recvtypes, test_comm, hs, isend, 0);
+        /* else if (itest == 1) */
+        /* { */
+        /*     hs = true; */
+        /*     isend = true; */
+        /*     ret = pio_swapm(ntasks, my_rank, sbuf,  sendcounts, sdispls, sendtypes, */
+        /*                     rbuf,  recvcounts, rdispls, recvtypes, test_comm, hs, isend, msg_cnt); */
+        /* } */
+        /* else if (itest == 2) */
+        /* { */
+        /*     hs = false; */
+        /*     isend = true; */
+        /*     ret = pio_swapm(ntasks, my_rank, sbuf, sendcounts, sdispls, sendtypes, */
+        /*                     rbuf, recvcounts, rdispls, recvtypes, test_comm, hs, isend, msg_cnt); */
+
+        /* } */
+        /* else if (itest == 3) */
+        /* { */
+        /*     hs = false; */
+        /*     isend = false; */
+        /*     ret = pio_swapm(ntasks, my_rank, sbuf, sendcounts, sdispls, sendtypes, */
+        /*                     rbuf, recvcounts, rdispls, recvtypes, test_comm, hs, isend, msg_cnt); */
+
+        /* } */
+        /* else if (itest == 4) */
+        /* { */
+        /*     hs = true; */
+        /*     isend = false; */
+        /*     ret = pio_swapm(ntasks, my_rank, sbuf,  sendcounts, sdispls, sendtypes, */
+        /*                     rbuf,  recvcounts, rdispls, recvtypes, test_comm, hs, isend, msg_cnt); */
+
+        /* } */
+
+        if (!my_rank)
+        {
+            gettimeofday(&t2, NULL);
+            printf("itest = %d Time in microseconds: %ld microseconds\n", itest,
+                   ((t2.tv_sec - t1.tv_sec) * 1000000L + t2.tv_usec) - t1.tv_usec);
+        }
+
+        /* Print results. */
+        if (!my_rank)
+        {
+            for (int e = 0; e < num_elem; e++)
+                printf("sbuf[%d] = %d", e, sbuf[e]);
+            for (int e = 0; e < num_elem; e++)
+                printf("rbuf[%d] = %d", e, rbuf[e]);
+        }
+    }
+
+    /* Test pio_fc_gather. In fact it does not work for msg_cnt > 0. */
```

Comment:
Yes, it will be. This is just the commented out code of my test for pio_fc_gather(), which turned out not to be needed, but looks very similar to what will be needed for pio_fc_gatherv(). After I get the swapm test working, I will uncomment this code and turn it into a test for pio_fc_gatherv().